flyio/caveats.md

```diff
@@ -0,0 +1,405 @@
+
+```
+  {
+    "type": "Apps",
+    "body": {
+      "apps": {
+        "0": "w"
+      }
+    }
+  },
+```
+
+
+### Other Resource Set Caveats
+
+There are other Resource Set Caveats. Unlike  the Apps Caveat, they use a string for the resource
+ID instead of a number, so the zero-valued key used as a wildcard must be the empty string instead
+of "0".  Other than that their behavior is very similar.  They allow an access if the resource
+being requested is in the caveat, and if the access being requested is a subset of the access
+allowed for that resource.  They are not relevant (return `ErrResourceUnspecified`) if the
+access request does not specify the corresponding resource.
+
+```
+  {
+    "type": "Volumes",
+    "body": {
+      "volumes": {
+        "volid": "w"
+      }
+    }
+  },
+```
+
+```
+  {
+    "type": "Machines",
+    "body": {
+      "machines": {
+        "machid1": "w",
+        "machid2": "w"
+      }
+    }
+  },
+```
+
+
+```
+  {
+    "type": "MachineFeatureSet",
+    "body": {
+      "features": {
+        "feat1": "w"
+      }
+    }
+  },
+```
+
+```
+  {
+    "type": "FeatureSet",
+    "body": {
+      "features": {
+        "feat1": "w"
+      }
+    }
+  },
+```
+
+```
+  {
+    "type": "Clusters",
+    "body": {
+      "clusters": {
+        "clust1": "w"
+      }
+    }
+  },
+```
+
+### IfPresent Caveat
+
+The IfPresent Caveat is a little bit different than other Caveats. It has an "if-then" part
+and an "else" part. The "if-then" part contains a list of Caveats. Access is allowed if
+all of the Caveats in the "if-then" part allow the access.  However, if all of the Caveats
+in the "if-then" part are not relevant (they all return `ErrResourceUnspecified`), then
+then instead the access is allowed if the requested action is a subset of the actions in
+the "else" clause. This is the only Caveat that behaves differently when some Caveats
+indicate that they are not relevant.
+
+Note: if ANY of the IfPresent Caveat's constituents are relevant, then ALL of the Caveats
+must allow the access.
+
+```
+  {
+    "type": "IfPresent",
+    "body": {
+      "ifs": [
+        {
+          "type": "Apps",
+          "body": {
+            "apps": {
+              "1234": "w"
+            }
+          }
+        }
+      ],
+      "else": "r"
+    }
+  }
+```
+
+### Mutations Caveat
+
+The Mutations Caveat restricts access to certain Mutations in the GraphQL API.
+An access request is allowed if it specifies a mutation that is one of the mutations
+listed in the Caveat.
+
+Mutation Caveats are not relevant (return `ErrResourceUnspecified`) if the
+access request does not specify a mutation.
+
+```
+  {
+    "type": "Mutations",
+    "body": {
+      "mutations": [
+        "mutation1",
+        "mutation2"
+      ]
+    }
+  },
+```
+
+### IsUser Caveat
+
+Not sure. To quote: `IsUser is mostyly metadata and plays no role in access validation.`
```

Comment:
This was from before we had attestations. We should be replacing this with 

https://github.com/superfly/macaroon/blob/d640a69989e6196f2c663b09cfaf4a8d11c6e236/auth/caveats.go#L200

Reply:
updated this, and added documentation on the "auth" caveats and access control checks on third party discharge tokens.